> I would also add `stacklevel=3` in calls of `warnings.warn()` and change tests to test it:
> 
> ```python
> with self.assertWarns(DeprecationWarning) as w:
>     Foo('test_func').run()
> self.assertIn('It is deprecated to return a value!=None', w.warnings[0].message)
> self.assertIn('test_func', w.warnings[0].message)
> self.assertEqual(w.warnings[0].filename, __file__)
> ```

When I used this idea, it works in regular test case, but when trying it with the async_case.py, I get:

```
     self.assertIn(w.warnings[0].filename, __file__)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: '/Users/ak/opensource/cpython/Lib/unittest/async_case.py' not found in '/Users/ak/opensource/cpython/Lib/unittest/test/test_async_case.py'
```

I will look more into it, just posting here in case you know why it doesn't work in the same way...

--- a/Lib/unittest/async_case.py
+++ b/Lib/unittest/async_case.py
@@ -65,7 +65,7 @@ class IsolatedAsyncioTestCase(TestCase):
     def _callTestMethod(self, method):
         if self._callMaybeAsync(method) is not None:
             warnings.warn(f'It is deprecated to return a value!=None from a '
-                          f'test case ({method})', DeprecationWarning)
+                          f'test case ({method})', DeprecationWarning, stacklevel=4)
 
     def _callTearDown(self):
         self._callAsync(self.asyncTearDown)

--- a/Lib/unittest/case.py
+++ b/Lib/unittest/case.py
@@ -548,7 +548,7 @@ class TestCase(object):
     def _callTestMethod(self, method):
         if method() is not None:
             warnings.warn(f'It is deprecated to return a value!=None from a '
-                          f'test case ({method})', DeprecationWarning)
+                          f'test case ({method})', DeprecationWarning, stacklevel=3)
 
     def _callTearDown(self):
         self.tearDown()

--- a/Lib/unittest/test/test_async_case.py
+++ b/Lib/unittest/test/test_async_case.py
@@ -172,9 +172,16 @@ class TestAsyncCase(unittest.TestCase):
         class Test(unittest.IsolatedAsyncioTestCase):
             async def test(self):
                 return 1
+            async def test2(self):
+                yield 1
 
         with self.assertWarns(DeprecationWarning) as w:
             Test('test').run()
+            self.assertIn('It is deprecated to return a value!=None', w.warnings[0].message.args[0])
+            self.assertIn('test', w.warnings[0].message.args[0])
+            self.assertIn(w.warnings[0].filename, __file__)
+        with self.assertWarns(DeprecationWarning) as w:
+            Test('test2').run()
             self.assertIn('It is deprecated to return a value!=None', w.warnings[0].message.args[0])
             self.assertIn('test', w.warnings[0].message.args[0])
             self.assertIn(w.warnings[0].filename, __file__)

--- a/Lib/unittest/test/test_async_case.py
+++ b/Lib/unittest/test/test_async_case.py
@@ -170,21 +170,22 @@ class TestAsyncCase(unittest.TestCase):
     def test_deprecation_of_return_val_from_test(self):
         # Issue 41322 - deprecate return of value!=None from a test
         class Test(unittest.IsolatedAsyncioTestCase):
-            async def test(self):
+            async def test1(self):
                 return 1
             async def test2(self):
                 yield 1
 
         with self.assertWarns(DeprecationWarning) as w:
-            Test('test').run()
-            self.assertIn('It is deprecated to return a value!=None', w.warnings[0].message.args[0])
-            self.assertIn('test', w.warnings[0].message.args[0])
-            self.assertIn(w.warnings[0].filename, __file__)
+            Test('test1').run()
+        self.assertIn('It is deprecated to return a value!=None', str(w.warnings[0].message))
+        self.assertIn('test1', str(w.warnings[0].message))
+        self.assertEquals(w.warnings[0].filename, __file__)
+
         with self.assertWarns(DeprecationWarning) as w:
             Test('test2').run()
-            self.assertIn('It is deprecated to return a value!=None', w.warnings[0].message.args[0])
-            self.assertIn('test', w.warnings[0].message.args[0])
-            self.assertIn(w.warnings[0].filename, __file__)
+        self.assertIn('It is deprecated to return a value!=None', str(w.warnings[0].message))
+        self.assertIn('test2', str(w.warnings[0].message))
+        self.assertEquals(w.warnings[0].filename, __file__)
 
     def test_cleanups_interleave_order(self):
         events = []

--- a/Lib/unittest/test/test_async_case.py
+++ b/Lib/unittest/test/test_async_case.py
@@ -173,9 +173,11 @@ class TestAsyncCase(unittest.TestCase):
             async def test(self):
                 return 1
 
-        with self.assertWarnsRegex(DeprecationWarning,
-                                   'It is deprecated to return a value!=None'):
+        with self.assertWarns(DeprecationWarning) as w:
             Test('test').run()
+            self.assertIn('It is deprecated to return a value!=None', w.warnings[0].message.args[0])
+            self.assertIn('test', w.warnings[0].message.args[0])
+            self.assertIn(w.warnings[0].filename, __file__)
 
     def test_cleanups_interleave_order(self):
         events = []

--- a/Lib/unittest/test/test_case.py
+++ b/Lib/unittest/test/test_case.py
@@ -309,21 +309,22 @@ class Test_TestCase(unittest.TestCase, TestEquality, TestHashing):
     def test_deprecation_of_return_val_from_test(self):
         # Issue 41322 - deprecate return of value!=None from a test
         class Foo(unittest.TestCase):
-            def test(self):
+            def test1(self):
                 return 1
             def test2(self):
                 yield 1
 
         with self.assertWarns(DeprecationWarning) as w:
-            Foo('test').run()
-            self.assertIn('It is deprecated to return a value!=None', w.warnings[0].message.args[0])
-            self.assertIn('test', w.warnings[0].message.args[0])
-            self.assertIn(w.warnings[0].filename, __file__)
+            Foo('test1').run()
+        self.assertIn('It is deprecated to return a value!=None', str(w.warnings[0].message))
+        self.assertIn('test1', str(w.warnings[0].message))
+        self.assertEquals(w.warnings[0].filename, __file__)
+
         with self.assertWarns(DeprecationWarning) as w:
             Foo('test2').run()
-            self.assertIn('It is deprecated to return a value!=None', w.warnings[0].message.args[0])
-            self.assertIn('test2', w.warnings[0].message.args[0])
-            self.assertEqual(w.warnings[0].filename, __file__)
+        self.assertIn('It is deprecated to return a value!=None', str(w.warnings[0].message))
+        self.assertIn('test2', str(w.warnings[0].message))
+        self.assertEquals(w.warnings[0].filename, __file__)
 
     def _check_call_order__subtests(self, result, events, expected_events):
         class Foo(Test.LoggingTestCase):

--- a/Lib/unittest/test/test_case.py
+++ b/Lib/unittest/test/test_case.py
@@ -314,12 +314,16 @@ class Test_TestCase(unittest.TestCase, TestEquality, TestHashing):
             def test2(self):
                 yield 1
 
-        with self.assertWarnsRegex(DeprecationWarning,
-                                   'It is deprecated to return a value!=None'):
+        with self.assertWarns(DeprecationWarning) as w:
             Foo('test').run()
-        with self.assertWarnsRegex(DeprecationWarning,
-                                   'It is deprecated to return a value!=None'):
+            self.assertIn('It is deprecated to return a value!=None', w.warnings[0].message.args[0])
+            self.assertIn('test', w.warnings[0].message.args[0])
+            self.assertIn(w.warnings[0].filename, __file__)
+        with self.assertWarns(DeprecationWarning) as w:
             Foo('test2').run()
+            self.assertIn('It is deprecated to return a value!=None', w.warnings[0].message.args[0])
+            self.assertIn('test2', w.warnings[0].message.args[0])
+            self.assertEqual(w.warnings[0].filename, __file__)
 
     def _check_call_order__subtests(self, result, events, expected_events):
         class Foo(Test.LoggingTestCase):

